### PR TITLE
Workhop instructions

### PR DESCRIPTION
**Pull request by rv0lt**

### CODE REVIEW

File: README.md
Comment (aishling-scilifelab) [inline README.md]: I don't think we can assume that the attendees will have git installed and I think getting them to install git (especially if they are Windows users) will add a bunch of faff. 
Also, many won't have a nice markdown renderer installed locally but it will render nicely on GitHub 

Can we just ask them to download the files under data using the download button?

File: data/example_directory_2/sub_directory_2/example_file_3.txt
Comment (aishling-scilifelab) [inline data/example_directory_2/sub_directory_2/example_file_3.txt:1]: I love this example file 😆

File: docs/1_installation.md
Comment (aishling-scilifelab) [inline docs/1_installation.md]: Super minor but probably don't need that full stop/period after Windows. 

```suggestion
They differ slightly between Mac or Windows so follow the section you need. Once installed, verify it by running:
```

File: docs/1_installation.md
Comment (aishling-scilifelab) [inline docs/1_installation.md:38]: Do we have any common issues during installation that we want to highlight the fix for?

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md]: The default accessibility alternate text still there - could add a brief description of the image.

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md]: Grammar 

```suggestion
 - [ ] TASK: Find the email and create an account and login using your information. Note that the password will not be shown while you are typing it.
```

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md]: 2 things: 
1) Might also be good to note that this doesn't automatically create an account for the PI (nor does it care whether a DDS a/c does exist with that address)
2) Capitalisation 
```suggestion
> The email specified in the option `--principal-investigator` does not receive any emails; it’s only for information purposes at this time.
```

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md]: Accessible image description missing

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md:54]: Did we want to do anything here about the non-required options (researcher, owner and non-sensitive)?

I'm guessing there will be some people who have never used DDS but we will also have people who do already use it so might be good to try and make them aware of some options they don't already use.

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md:73]: This command won't work on Windows. 

Need to use `dir`

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md]: Same comment as above: it might be cool to pick 1 or 2 of the options (especially if there are some that are more useful) and have users try them out too

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md:107]: What does 'interactively' mean here?

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md]:
```suggestion
However, you can view all files and directories in a project as a tree structure, using the `--tree` option.
```

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md]: Typo 

```suggestion
 - [ ] TASK: Verify that all the data has being uploaded successfully
```

Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md]: Do we have any tips for users (possibly not DDS commands but Linux or Windows CMD commands) for sanity checking the upload given that some uploads might have 10,000s of files? 
For example, commands to count the number of files in a dir (e.g. `ls . | wc -l`) or get the total size of all files in the directory (e.g. `du -hs ./*`)

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md]: Might be interesting to show them (even if we don't get them to test it) what researchers see if they try to download data when the project is not set to Available.

Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md]: Might also be good to mention that putting the project into Available starts a countdown in number of days

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md]: Might be good to note what the difference is: when should they pick one over the other?

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md]: I think we should tell them about considering zipping before upload (because we want them to do that)

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md]: I get what you mean here but for some reason find this sentence hard to parse. Maybe could say:
_If you have another account on Test DDS, you can invite that email as a researcher_
That said, 1) couldn't they just use plus addressing to instantly invite themselves as a researcher anyway? 2) how does it work when you want to set up authentication for 2 different accounts on the same computer? Presumably they'd have to end/delete the stored session token and then re-auth with the other account's credentials. Might be good to mention that

**Review by aishling-scilifelab — CHANGES_REQUESTED** [review]
This looks good!

Unless we're covering it elsewhere, I think we should demonstrate the expected behaviour when both an upload and download are interrupted.

File: README.md
Comment (valyo) [reply README.md]: We can mention both git and straight download (for those without git installed). Git is usually part of the reproducible research workflows and it will add an educational effect to our material, I believe.

File: README.md
Comment (valyo) [inline README.md]:
```suggestion
for SciLifeLab Unit Users (Autumn 2024).
```

File: docs/1_installation.md
Comment (valyo) [inline docs/1_installation.md]:
```suggestion
## Using the testing instance
```

File: docs/1_installation.md
Comment (valyo) [inline docs/1_installation.md]:
```suggestion
By default the client is using the production instance with URL

> `https://delivery.scilifelab.se`

You can see that URL in the output of any command you run.

```

File: docs/1_installation.md
Comment (valyo) [reply docs/1_installation.md:38]: Joanna had an issue with missing `python3-devel ` package on Fedora linux, so I guess we can add that people with non Ubuntu/Debian based distros can possibly need to do smth like 
`sudo yum install python3-devel`

File: docs/2_registration_project_access_session.md
Comment (rv0lt) [reply docs/2_registration_project_access_session.md]: It is just so they could test adding emails as a researcher, I don't intend on them to log in again. Because that could be too much mess.

File: docs/2_registration_project_access_session.md
Comment (rv0lt) [reply docs/2_registration_project_access_session.md]: No difference at all, just two commands that do the same. The second one can be used to add any other use if they change the role. But for this purposes it is the same.

File: docs/2_registration_project_access_session.md
Comment (rv0lt) [reply docs/2_registration_project_access_session.md]: Should be mentioned in project life cycle

Comment (rv0lt) [reply docs/2_registration_project_access_session.md]: It's not so mysterious, the client just stops and tells them to wait until the project is available (has happened a lot of times to me while locally testing).

Also, they get an email when it is released. So they know about it.

File: docs/2_registration_project_access_session.md
Comment (rv0lt) [reply docs/2_registration_project_access_session.md:107]: That the command prompts them again to specify a subdirectory to navigate through the files

File: docs/2_registration_project_access_session.md
Comment (rv0lt) [reply docs/2_registration_project_access_session.md]: Some of the other flags from put are way more advanced.
I prefer to keep this tutorial simple. And during the session explain the documentation and the options.

File: docs/2_registration_project_access_session.md
Comment (rv0lt) [reply docs/2_registration_project_access_session.md:54]: I think those more advanced questions, we can mention them live. Specially if they ask for them.

It will be difficult to account for all the specific options each command has, but all the commands explained have links to the docs, where all the flags are explained. And we can project the documentation while we explain the commands.

File: docs/1_installation.md
Comment (rv0lt) [reply docs/1_installation.md:38]: What do you mean?

File: README.md
Comment (rv0lt) [reply README.md]: I rephrase it. 
But better if they download the whole repository as ZIP, if they download individual files, they can miss some of them and break the structure shown in the examples.

But anyway, it is not like it is asynchronous. We will be showing them in the screen this steps

File: docs/2_registration_project_access_session.md
Comment (rv0lt) [reply docs/2_registration_project_access_session.md]: Actually, you reminded me of something. 
One of these commands shows the size as Kibibytes, and our system stores the data as kilobytes (or the other way around, don't remember exactly). So when big uploads, there is a big difference on size if someone manually checks like that. Need to check it

File: docs/2_registration_project_access_session.md
Comment (valyo) [reply docs/2_registration_project_access_session.md]: I agree with both of your points here, maybe we can have a task during the session where they look at the options and try to figure them out and/or raise questions about them

File: docs/2_registration_project_access_session.md
Comment (valyo) [inline docs/2_registration_project_access_session.md]:
```suggestion
> NOTE: As of today, DDS has performance issues and upload failures when uploading many small files. We are working on fixing it, but on the meantime, we recommend you to archive (with ZIP or RAR, no compression required) files when the delivered data consists of hundreds or thousands of small files).
```

File: docs/2_registration_project_access_session.md
Comment (valyo) [reply docs/2_registration_project_access_session.md]: Isn't the first command intended for existing accounts, i.e. it should fail if there is no account in the system with the specified email address (not sure if that's the behaviour, we should check it)

File: README.md
Comment (aishling-scilifelab) [reply README.md]: Totally get the point for the data directory (and maintaining that structure) but most of these people (likely) won't have an IDE with a markdown renderer so we still want them to look at the markdown instructions in GitHub (where it all looks nice and well-formatted). 

Could you zip the data files and commit that zipped file? Then they can simply download that.

File: docs/1_installation.md
Comment (aishling-scilifelab) [reply docs/1_installation.md:38]: As in, as well as assuming installation all goes well (especially as these users are supposed to support their data recipients in installing), it might be nice to talk about errors which can be encountered at this point (e.g. that have been previously reported to us).

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [reply docs/2_registration_project_access_session.md:54]: I think if we're going to talk through it, we should have it written down here (possibly in a box or section labelled Optional). Given that there will be ~15 people, they likely won't remain perfectly in sync with each other and there will be people who won't be able to keep up with the live presentation aspect

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [reply docs/2_registration_project_access_session.md]: Same as previous comment - realistically not all the attendees will be able to keep pace (and some will already be advanced users so will be familiar with these commands). If we are going to talk about the options then I think we should add a little section on that (but possibly labelling it optional)

File: README.md
Comment (rv0lt) [reply README.md]: I mean yes. But they can already download as a zip the repo. 
And that doesn't imply they have to open the markdown files raw, like they can still follow the docs through the browser. I don't see a problem here.

File: docs/1_installation.md
Comment (rv0lt) [reply docs/1_installation.md:38]: Weren't we asking them to have it already installed? I don't think we have written down a historical of previous reported cases. It can depend on a lot of things, I cannot think of concrete examples at this point

File: docs/2_registration_project_access_session.md
Comment (rv0lt) [reply docs/2_registration_project_access_session.md:54]: I will mention it as a note, okay

File: docs/2_registration_project_access_session.md
Comment (rv0lt) [reply docs/2_registration_project_access_session.md]: The optional flags for this command can be tricky, I can mention as a note that there are other options they can look up in the documentation (which they already have linked) but explain them can confuse more

File: docs/1_installation.md
Comment (valyo) [reply docs/1_installation.md:38]: That Fedora "issue" is the only one I can think of, and it is worth it mentioning it

Comment (rv0lt) [reply docs/1_installation.md:38]: I have searched more deeply, and I cannot find any document talking about installation errors

File: docs/2_registration_project_access_session.md
Comment (rv0lt) [reply docs/2_registration_project_access_session.md]: I added a small explanation as a note

File: docs/2_registration_project_access_session.md
Comment (rv0lt) [reply docs/2_registration_project_access_session.md]: Clarified that they do the same along with a reminder that they can also invite researcher when creating project

File: docs/1_installation.md
Comment (valyo) [reply docs/1_installation.md:38]: I don't think there's a document anywhere. This was Joanna trying to install the CLI on her Fedora computer and getting errors since Fedora doesn't seem to have that package installed by default. I don't think you'll find any other sources about such issue because there're no so many people around the world that tried to install our CLI and discussed the issue. The idea is that we mention it as a potential guide for people using other distros, since we cannot cover all of them, of cource.

File: docs/2_registration_project_access_session.md
Comment (valyo) [inline docs/2_registration_project_access_session.md:122]:
```suggestion
for giving access to an already existing account, or
```

File: docs/2_registration_project_access_session.md
Comment (valyo) [inline docs/2_registration_project_access_session.md:127]:
```suggestion

for sending an invitation for a new account, which will automatically obtain access to that specific project.

```

File: docs/2_registration_project_access_session.md
Comment (valyo) [inline docs/2_registration_project_access_session.md]:
```suggestion
>As mentioned above, it is possible to automatically invite researchers when creating  a project. \
```

**Review by valyo — CHANGES_REQUESTED** [review]
Let's have all the suggestion applied and can have another iteration.

File: docs/2_registration_project_access_session.md
Comment (valyo) [inline docs/2_registration_project_access_session.md]:
```suggestion
> There are a number of other optional flags, which can be found on the documentation link just above.
```

File: docs/1_installation.md
Comment (aishling-scilifelab) [reply docs/1_installation.md:38]: @valyo - can you summarise what Joanna had to install and I'll add some text in here?

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [inline docs/2_registration_project_access_session.md]: @valyo - given this suggestion, would you disagree with the text (see discussion immediately above this one) that says the commands are the same? 
Edit: I can see from below that yes you disagree there is no difference.

File: docs/2_registration_project_access_session.md
Comment (aishling-scilifelab) [reply docs/2_registration_project_access_session.md]: I still don't really understand this one.

Comment (valyo) [reply docs/2_registration_project_access_session.md]: Yes, this is one of the places where the flow gets confusing to me. Maybe we should rearrange things, because the section title is "Release project and download data" but it includes giving access to researchers. Probably cleaner if we keep things to unit personnel here, i.e. release project and download data

File: docs/2_registration_project_access_session.md
Comment (valyo) [reply docs/2_registration_project_access_session.md]: The end result is the same in terms that a researcher account gets access to a project, but the commands are for different scenarios. So I wouldn't say the commands are the same. Maybe we don't even need to say explicitly that they are the same or different. I will thy to find a solution

File: docs/1_installation.md
Comment (aishling-scilifelab) [reply docs/1_installation.md:38]: Actually, it still isn't working for Joanna so let's come back and add instructions for Fedora later.

**Comment by aishling-scilifelab** [comment]
@valyo - I have left 2 things unresolved: 1) Installation on Fedora and 2) have left Alvaro's text stating there is no difference between those 2 commands (because Ina's going to double check that later in the week).

File: docs/2_registration_project_access_session.md
Comment (i-oden) [reply docs/2_registration_project_access_session.md]: So I checked the code a bit quickly now and the `dds project access grant` command calls the `AddUser` resource / endpoint, which is the same one that `dds user add` calls, so there's no difference in functionality. I guess the idea is that they use `dds user add` (which was the only command from the beginning) as a default when they're inviting new accounts and that they use `dds project access grant` when they want to give an existing account access, but they could technically use both for both purposes. Not sure atm what's best but I feel like I personally would just want to know about one command first ("the intended way") and then possibly have a note somewhere where it says 'you could also use xxx', maybe including both here can cause some confusion / more questions.

File: docs/2_registration_project_access_session.md
Comment (valyo) [inline docs/2_registration_project_access_session.md]:
```suggestion
> There is no practical difference between these two commands for this use case. As mentioned above, it is 
> also possible to automatically invite researchers when creating a project. `dds project create (...) --researcher [Email address]`
```

File: docs/2_registration_project_access_session.md
Comment (valyo) [reply docs/2_registration_project_access_session.md]: I believed that `dds project access grant --project "<Project ID>" --email "<email>"` fail if the specified email doesn't already have an account registered, but it seems I was wrong. Just checked it and it turns out it sends an invitation if there's no account. 
We can leave it as it is, with minor change in the text

File: docs/2_registration_project_access_session.md
Comment (i-oden) [reply docs/2_registration_project_access_session.md]: Let me know if I'm misinterpreting etc.

But basically this is how it works:
- If they want to be able to download with a researcher account then they'd need to invite themselves as a researcher to that project, and then that researcher email would need to register an account. If that email has already been invited to the system and they have registered a researcher account previously, then they can invite that email to a project, and they then automatically have access to download, they wouldn't need to register again.
- In order to download with that researcher account, first they'd need to re-auth with the researcher credentials, and that would replace the existing local token (that belonged to the authenticated unit user account). 

So I think the "you don't need to register again" should be removed/moved somewhere else? If we want to have this optional suggestion here we could rephrase it to something like this, perhaps? (But we may want to include steps in that case)
> - [ ] Optional: If you have another email address, use your Unit Admin / Personnel account to invite that email as a researcher. Once you have registered the "researcher" account, authenticate using the researcher account credentials and use the `dds data get` command as described above to download the data.

Comment (valyo) [reply docs/2_registration_project_access_session.md]: We should either change the text as Ina suggests, or don't mess things with Researcher account here. 
Maybe the best is to make this change now, and later on decide if it should go to another section or removed all together.

File: docs/2_registration_project_access_session.md
Comment (valyo) [inline docs/2_registration_project_access_session.md]:
```suggestion
 - [ ] Optional: If you have another email address, use your Unit Admin / Personnel account to invite that email as a researcher. Once you have registered the "researcher" account, authenticate using the researcher account credentials and use the dds data get command as described above to download the data.
```

File: docs/2_registration_project_access_session.md
Comment (valyo) [inline docs/2_registration_project_access_session.md]: This one should be in the next section, IMO. 
Since it is too messy to suggest the change here, I will do it locally and commit it

**Comment by aishling-scilifelab** [comment]
@valyo can we get this one merged in, even if it's not perfect? Now that Alvaro is back especially it's easier if we can make little branches off main

**Review by valyo — CHANGES_REQUESTED** [review]
Let's get all suggestions applied or denied, so that we can have clean view of the current material.

**Comment by aishling-scilifelab** [comment]
@valyo can you apply those requested changes (I think you have the clearest picture of that ATM)?

**Review by valyo — APPROVED** [review]

**Comment by valyo** [comment]
@aishling-scilifelab I applied the changes and dismissed my changes request, now it's yours remaining and we can merge it

**Review by aishling-scilifelab — APPROVED** [review]

**Comment by valyo** [comment]
I committed some changes that I already had, but it was probably stupid since the PR already has approvals. Please, take a look and merge, and then wi can continue in new PRs

**Review by aishling-scilifelab — APPROVED** [review]

**Pull request merged**
